docs/source/release_roadmap.org

```diff
@@ -0,0 +1,363 @@
+#+TITLE: Recommended Road
+#+AUTHOR: Aleksandr Petrosyan
+#+DATE: 2023-08-13
+
+* State of the project
+
+Iroha v2 is a capable ledger with notable implemented subroutines that
+allow it to be deployed successfully as a private blockchain ledger:
+1. It has relatively good performance characteristics regarding
+   throughput (post RC9)
+2. It has good performance characteristics regarding transaction
+   finality and latency (post RC19 also query latency).
+3. It has support for Turing complete permission validation, able to
+   enforce arbitrarily complex logic on-chain
+4. It has support for a small interpreted VM called the Iroha Special
+   Instructions which allow small-footprint transactions
+5. It has support for Turing complete smartcontracts living in a
+   sandbox, with a full array of events which can trigger these
+   behaviours.
+6. The ledger has three SDKs within the JVM, JavaScript and Python
+   ecosystems, allowing the users to easily adapt the Iroha node to
+   their requirements. 
+
+
+The ledger has been demonstrated to work during the ADB CSD/RTGS
+linkages project, and is slated to be used in two projects internally
+by Soramitsu which only utilise the private blockchain ledger
+functionality.
+
+The ledger can support un-sharded monolithic blockchains, with limited
+export capability (unlike Iroha v1, v2 uses an internal representation
+and doesn't use a database). Its consensus was designed with Byzantine
+fault tolerance in mind, however, the algorithm -- =sumeragi= has some
+intrinsic flaws outlined in [[https://arxiv.org/pdf/1707.01873.pdf][an IBM Research article]], which is notable
+because IBM research originally designed the B-Chain algorithm which
+serves as a prototype for =sumeragi=.
+
+The blockchain ledger is designed to support a plug-able (and
+upgrade-able) permission validator modules, allowing the blockchain
+invariants to evolve and change with the operation of the blockchain.
+It is capable of resolving simple soft-forks, and has tooling to
+quickly restart a blockchain by squashing all of the information and
+history into a single genesis block. However, it does not retain
+enough flexibility to be used as a single ledger for both private and
+public blockchains. And the following issues have been identified.
+
+1. The more efficient ISI-based language is not usable. The
+   expressions are not sufficiently expressive.
+
+   While this can be somewhat mitigated by the introduction of the
+   DSL, the main expressiveness-boosting facilities of ISI are not
+   usable, and thus only add overhead. A thorough exploration of a
+   restructuring of the Iroha Instruction Set Architecture is needed
+   and shall be published as an RFC in August 2023.
+
+2. WASM is a sub-optimal medium for storage of executable logic. It
+   alone is responsible for 80% of the complexity present in the code
+   base and around 95% of technical debt is related to some mitigation
+   of a shortcoming of WASM in the blockchain.
```

Comment:
tl;dr it also limits SDK usability

referring to SDKs, a developer willing to use smart contracts is basically forced to write ones in Rust since compiling almost everything else to WASM is experimental/prototypical/highly challenging
moving that to the side of SDKs is an overkill at this point

Reply:
ah, I see it basically mentioned later, nvm

Reply:
@Mingela hmm, i don’t see how picking other runtime will stop limiting SDK devs. If we will pick lua then sevs have to write contracts in lua. If jvm them write contract in jvm based language and so on…
Can you explain it a bit further?

Reply:
Well, this will be expanded upon in the WASM-specific document. The main issue with WASM specifically is that it is tailored to a web-based use, which has some countervailing tendencies to what we see in general. There's no pressure to stabilise modules, hence no dynamic linkage. It is meant to transport low-level compiled instructions thus no or non-existent run-time support is present. WASM takes up more RAM. WASM must be compiled. WASM is not backwards compatible and WASM must be AOT compiled and then JIT-ted for the particular run-time.  

With a different standard we get a bit more freedom, because the different standard is designed for specific uses. The JVM is perhaps the closest point of comparison. It too has a limited number of supported languages that can target it, but a much wider coverage. Because most JVM languages assume a host-side garbage collector, one has automatic memory management without either the borrow checker, or having to embed GC. Java is smaller for the same source code, because more of the data is embedded in the runtime. 

The runtime itself can be large, but because of late binding the actual memory usage is smaller. Plus we don't have to come up with a dynamic linkage mechanism, we have one. We also have functional multithreading, and 64-bit support. Another important aspect for why JVM would be better is the guaranteed backwards compatibility. Java does not make breaking changes, and you don't have to keep an older version of a runtime in order to preserve compatibility. This means that instead of a complex system of host-side runtime selection, we can just upgrade to the newer version. 

Thus with Java we get a slightly less complex language for smartcontracts that can do slightly more, with the added benefit of being a huge ecosystem tailored to running application logic (and not DOM manipulation), that is significantly more mature. WASM is not the worst possible standard, but we can alleviate a lot of the problems. 

AND THEN, we get to interpreted languages, which have lightweight syntax, don't allow you to do what's not intended, can live concurrently and allow for a lot of transparency. Once I give you the WASM blob, you have a whole song and dance to verify that it belongs to the source code that I gave you. Most people wouldn't bother and that discourages audits. At the same time, saying that the smartcontract is an actual source snippet that is interpreted in the blockchain, tells you a lot more about the logic of the smartcontract and gives you a lot more trust. 

Because for an interpreted language we must expose the FFI bindings, we could then just say "you know what, we don't care, as long as it conforms to some very simple protocol, use Python if that rocks your boat". Then the language becomes an implementation detail for each blockchain, and we're not confined to _one_ choice.